### DISCUSSION TITLE
AppStorage で不足な場合は Defaults の利用を検討する。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#23] [Ideas]
## Overview

SwiftUI 2.0 からはデータを保持する仕組みとして [@AppStorage](https://developer.apple.com/documentation/swiftui/appstorage) が導入された。これは UserDefaults に対するラッパー的なもので、アプリの設定値などを保存する際に利用できる。

以下は `Toggle` の選択値（`true` or `false`）を`"isAutomaticallyUpdate"`というキー名で永続化するシンプルな例となる。

```swift
struct ContentView: View {
    @AppStorage("isAutomaticallyUpdate") var isAutomaticallyUpdate = false
    
    var body: some View {
        Toggle("Automatically update", isOn: $isAutomaticallyUpdate)
            .padding()
    }
}
```

<img width="418" alt="image" src="https://user-images.githubusercontent.com/2990285/162117721-0e546071-6a20-4c63-aff4-0d9d31c2d965.png">

これはシンプルで強力に見えるが、現状の `@AppStorage` には多くの制限が存在する。

不足を感じる場合は OSS として開発されている [sindresorhus/Defaults](https://github.com/sindresorhus/Defaults) の利用を検討するとよい。

## Discussion

まず、`@AppStorage`では扱えるデータ型が限られている。

- `String`
- `Data`
- `Int`
- `URL`
- `Double`
- `Bool`
- `RawRepresentable (Value == Int)`
- `RawRepresentable (Value == String)`

`Data`がサポートされているので実質的には任意のデータを扱えると言えるが、自前で変換する必要があるためタイプセーフなコードではなくなる。

次に Optional な `RawRepresentable` でデフォルト値（`wrappedValue`）を与えるイニシャライザは定義されていない。

```swift
enum SelectItem: Int {
    case all
    case star
}

struct ContentView: View {
    
    // ✅ This is work.
    @State var selected: SelectItem? = .all
    
    // 🚫 Compile error.
    @AppStorage("selected") var selected: SelectItem? = .all


    var body: some View {
        NavigationView {
            List {
                // 💡 `selection` requires `Binding<V?>` type.
                NavigationLink(tag: SelectItem.all, selection: $selected, destination: { Text("Hello") }) {
                    Text("All")
                }
                NavigationLink(tag: SelectItem.star, selection: $selected, destination: { Text("Goodbye") }) {
                    Text("Star")
                }
            }
        }
    }
}
```

このコード例において `NavigationLink` の `selection` は `Binding<V?>` 型を要求するため、`SelectItem` 型に変更すると今度はそちらでコンパイルエラーになり、それを解決するためには自前の Binding ラッパーを書く必要がある。

その他、離れた View から参照するためには Binding を引き渡す必要があったり、2回目移行の書き込みをキャッシュ（しているように見える挙動を）するなど、色々と面倒なことも多い。

[sindresorhus/Defaults](https://github.com/sindresorhus/Defaults) の README には、[@AppStorage に対するメリット](https://github.com/sindresorhus/Defaults#benefits-over-appstorage)も明記されているので、参考にするとよい。

## Note

タブやリストの選択状態など、ウィンドウ・シーン間で共有したくない値については [@SceneStorage](https://developer.apple.com/documentation/swiftui/scenestorage) を利用すること。

## Example

以下のアプリでは [@AppStorage のみを使用](https://github.com/YusukeHosonuma/E2DC/blob/main/Root/Sources/Root/View/RootView.swift#L15) しています。
https://github.com/YusukeHosonuma/E2DC

以下のアプリでは [Defaults のみを使用](https://github.com/YusukeHosonuma/Swift-Evolution-Browser/blob/main/Root/Sources/Root/View/RootView.swift#L23) しています。

ごくシンプルなケースで済む場合は、標準である `@AppStorage` で済ますのも手かと思います。

## Reference

- [@AppStorage](https://developer.apple.com/documentation/swiftui/appstorage) 
- [SceneStorage](https://developer.apple.com/documentation/swiftui/scenestorage)
- [sindresorhus/Defaults](https://github.com/sindresorhus/Defaults)

**Comment by uhooi** [comment]
UserDefaults のラッパーは簡単に作れるので、できる限りサードパーティ製のライブラリに依存しないのが好みです…！
